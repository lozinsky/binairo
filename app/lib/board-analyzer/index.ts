export * from './analyze-board-by-kind';
export * from './analyze-board';
