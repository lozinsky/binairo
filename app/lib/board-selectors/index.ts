export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
