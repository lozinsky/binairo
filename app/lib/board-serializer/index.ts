export * from './board-serializer-collection';
