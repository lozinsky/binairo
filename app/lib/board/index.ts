export * from './board-cell';
export * from './board-line';
export * from './board';
