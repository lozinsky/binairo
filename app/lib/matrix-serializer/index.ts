export * from './matrix-selection-serializer-collection';
