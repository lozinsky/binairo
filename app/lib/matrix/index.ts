export * from './matrix-line';
export * from './matrix-rotation';
export * from './matrix-selection';
export * from './matrix';
