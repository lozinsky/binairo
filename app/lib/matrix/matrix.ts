import type { MatrixLine, MatrixLineCell } from './matrix-line';

export interface Matrix<L extends MatrixLine<unknown>> {
  at(index: number): L | undefined;
  entries(): Iterable<[number, L]>;
}

export type MatrixCell<M> = M extends Matrix<infer L> ? MatrixLineCell<L> : never;
