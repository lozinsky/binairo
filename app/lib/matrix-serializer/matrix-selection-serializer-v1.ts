import { isMatrixSelectionPositionPositions, MatrixSelection } from '~/lib/matrix';
import { decodeURLBase64, encodeURLBase64 } from '~/shared/url-base64';

import {
  type MatrixSelectionSerializer,
  MatrixSelectionSerializerDeserializeError,
} from './matrix-selection-serializer';

export class MatrixSelectionSerializerV1 implements MatrixSelectionSerializer {
  deserialize(data: string): MatrixSelection {
    let value: unknown;

    try {
      value = JSON.parse(decodeURLBase64(data));
    } catch (error) {
      throw new MatrixSelectionSerializerDeserializeError('Failed to deserialize matrix selection', { cause: error });
    }

    if (!isMatrixSelectionPositionPositions(value)) {
      throw new MatrixSelectionSerializerDeserializeError('Deserialized value is not a valid matrix selection');
    }

    return new MatrixSelection(value);
  }

  serialize(selection: MatrixSelection): string {
    return encodeURLBase64(JSON.stringify(selection.valueOf()));
  }
}
