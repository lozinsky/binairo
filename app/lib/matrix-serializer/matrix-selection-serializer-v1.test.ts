import { expect, test } from 'vitest';

import { MatrixSelection } from '~/lib/matrix';

import { MatrixSelectionSerializerV1 } from './matrix-selection-serializer-v1';

test.each([
  '',
  '-W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX1d',
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MX1d',
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX0',
])('throws on invalid data', (data) => {
  expect(() => new MatrixSelectionSerializerV1().deserialize(data)).toThrowErrorMatchingSnapshot();
});

test.each([
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX1d',
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX0seyJ4IjowLCJ5IjoyfSx7IngiOjIsInkiOjJ9XQ',
  'W3sieCI6MSwieSI6MX0seyJ4IjoxMCwieSI6MTB9LHsieCI6MTAwLCJ5IjoxMDB9LHsieCI6MTAwMCwieSI6MTAwMH0seyJ4IjoxMDAwMCwieSI6MTAwMDB9XQ',
])('returns matrix selection from string', (data) => {
  expect(new MatrixSelectionSerializerV1().deserialize(data)).toMatchSnapshot();
});

test.each([
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]),
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
    { x: 0, y: 2 },
    { x: 2, y: 2 },
  ]),
  new MatrixSelection([
    { x: 1, y: 1 },
    { x: 10, y: 10 },
    { x: 100, y: 100 },
    { x: 1000, y: 1000 },
    { x: 10000, y: 10000 },
  ]),
])('returns matrix selection string', (selection) => {
  expect(new MatrixSelectionSerializerV1().serialize(selection)).toMatchSnapshot();
});
