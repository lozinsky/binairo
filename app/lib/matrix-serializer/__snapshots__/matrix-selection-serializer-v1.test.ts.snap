// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns matrix selection from string 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`returns matrix selection from string 2`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 0,
    "y": 2,
  },
  {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`returns matrix selection from string 3`] = `
[
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 10,
    "y": 10,
  },
  {
    "x": 100,
    "y": 100,
  },
  {
    "x": 1000,
    "y": 1000,
  },
  {
    "x": 10000,
    "y": 10000,
  },
]
`;

exports[`returns matrix selection string 1`] = `"W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX1d"`;

exports[`returns matrix selection string 2`] = `"W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX0seyJ4IjowLCJ5IjoyfSx7IngiOjIsInkiOjJ9XQ"`;

exports[`returns matrix selection string 3`] = `"W3sieCI6MSwieSI6MX0seyJ4IjoxMCwieSI6MTB9LHsieCI6MTAwLCJ5IjoxMDB9LHsieCI6MTAwMCwieSI6MTAwMH0seyJ4IjoxMDAwMCwieSI6MTAwMDB9XQ"`;

exports[`throws on invalid data 1`] = `[MatrixSelectionSerializerDeserializeError: Failed to deserialize matrix selection]`;

exports[`throws on invalid data 2`] = `[MatrixSelectionSerializerDeserializeError: Failed to deserialize matrix selection]`;

exports[`throws on invalid data 3`] = `[MatrixSelectionSerializerDeserializeError: Deserialized value is not a valid matrix selection]`;

exports[`throws on invalid data 4`] = `[MatrixSelectionSerializerDeserializeError: Failed to deserialize matrix selection]`;
